test-tests/shared/src/test/scala/zio/test/environment/ClockSpec.scala

```diff
@@ -0,0 +1,157 @@
+            _     <- adjust(1.millis)
+            time2 <- currentTime(TimeUnit.NANOSECONDS)
+          } yield assert(time2 - time1, equalTo(1.millis.toNanos))
+        },
+        testM("adjust correctly advances currentDateTime") {
+          for {
+            time1 <- currentDateTime
+            _     <- adjust(1.millis)
+            time2 <- currentDateTime
+          } yield assert((time2.toInstant.toEpochMilli - time1.toInstant.toEpochMilli), equalTo(1L))
+        },
+        testM("adjust does not produce sleeps") {
+          adjust(1.millis) *> assertM(sleeps, isEmpty)
+        },
+        testM("setTime correctly sets nanotime") {
+          for {
+            _    <- setTime(1.millis)
+            time <- clock.nanoTime
+          } yield assert(time, equalTo(1.millis.toNanos))
+        },
+        testM("setTime correctly sets currentTime") {
+          for {
+            _    <- setTime(1.millis)
+            time <- currentTime(TimeUnit.NANOSECONDS)
+          } yield assert(time, equalTo(1.millis.toNanos))
+        },
+        testM("setTime correctly sets currentDateTime") {
+          for {
+            _    <- TestClock.setTime(1.millis)
+            time <- currentDateTime
+          } yield assert(time.toInstant.toEpochMilli, equalTo(1.millis.toMillis))
+        },
+        testM("setTime does not produce sleeps") {
+          for {
+            _      <- setTime(1.millis)
+            sleeps <- sleeps
+          } yield assert(sleeps, isEmpty)
+        },
+        testM("setTimeZone correctly sets timeZone") {
+          setTimeZone(ZoneId.of("UTC+10")) *>
+            assertM(timeZone, equalTo(ZoneId.of("UTC+10")))
+        },
+        testM("setTimeZone does not produce sleeps") {
+          setTimeZone(ZoneId.of("UTC+11")) *>
+            assertM(sleeps, isEmpty)
+        },
+        testM("timeout example from TestClock documentation works correctly") {
+          val example = for {
+            fiber  <- ZIO.sleep(5.minutes).timeout(1.minute).fork
+            _      <- TestClock.adjust(1.minute)
+            result <- fiber.join
+          } yield result == None
+          assertM(example, isTrue)
+        } @@ nonFlaky(100),
+        testM("recurrence example from TestClock documentation works correctly") {
+          val example = for {
+            q <- Queue.unbounded[Unit]
+            _ <- (q.offer(()).delay(60.minutes)).forever.fork
+            a <- q.poll.map(_.isEmpty)
+            _ <- TestClock.adjust(60.minutes)
+            b <- q.take.as(true)
+            c <- q.poll.map(_.isEmpty)
+            _ <- TestClock.adjust(60.minutes)
+            d <- q.take.as(true)
+            e <- q.poll.map(_.isEmpty)
+          } yield a && b && c && d && e
+          assertM(example, isTrue)
+        },
+        //todo fix - looks like fiberTime is 100x larger than it's supposed to be - conversion error somewhere?
+        testM("fiber time is not subject to race conditions") {
```

Comment:
The problem here is that since `zipPar` is joining back the fibers after each run the fiber time is being incremented on each run by 2 milliseconds, which over the course of 100 runs becomes 200 milliseconds, hence your off by a factor of 100 error. This is getting further complicated by the bug in `nonFlaky` I am fixing in #2013 which is causing the test to run for all 100 iterations instead of failing early.

You can fix this by keeping the clock time in sync with the fiber time.

```scala
testM("fiber time is not subject to race conditions") {
  for {
    _        <- adjust(2.millis)
    _        <- sleep(2.millis).zipPar(sleep(1.millis))
    result   <- fiberTime
    expected <- clock.currentTime(TimeUnit.MILLISECONDS)
  } yield assert(result.toMillis, equalTo(expected))
} @@ nonFlaky(100)
```

Reply:
as below I don't think this is testing the same thing anymore. The original test was pushing the clock time out to infinity, so it'd never be in sync with fiber time even the first time through. I think what it was trying to test is that with the clock set to infinity, fiber time still advances by the expected amount (2ms) each time

Reply:
The goal of this test is to check for a race condition in `ZipPar` with regard to the fiber time. As long as the clock time is equal to or greater than 2 milliseconds (ignoring repeated runs), both the left and right fibers in `zipPar` will wake up. There is a potential race condition that normally when child fibers are joined the `FiberRef` of the "last child" joined takes precedence. So there could be a situation where the left join takes places before the right join, leading to the right fiber time of 1 millisecond being inherited by the parent and leading to data corruption. We solved that by adding logic to `FiberRef` allowing to specify custom logic for joining and using a "greater duration wins" rule and this test is to prevent regressions in that. So in the new test the only way the result is equal to the expected value each time is if the fiber time is always incremented by 2 milliseconds, which is what we want.

There are definitely other way you could do this if you want but this seemed like a natural one to me.